### PR TITLE
github-copilot-intellij-agent: init at 1.2.8.2631

### PR DESCRIPTION
**Pull request by hacker1024**
###### Description of changes

This patches and packages the GitHub Copilot IntelliJ plugin's native component to use Nix libraries.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kouyk** [comment]
#191341 already exists, is it better to integrate or write the patch for copilot based on that so IDE updates won't break this?

**Comment by hacker1024** [comment]
Once that PR lands, a Copilot plugin derivation could be made that repackages the original plugin with the patched binary made in this derivation. This would allow for more flexibility, in case anyone does want to work with the agent alone.

**Comment by unitythemaker** [comment]
@hacker1024 I wanted to say thank you. I was being annoyed with Copilot not working, this one worked like a charm! It worked for GoLand, I didn't try but it should work for other IDEs too.

**Comment by Vinetos** [comment]
Works very well ! Big thanks !

File: pkgs/development/tools/github-copilot-intellij-agent/default.nix [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    fix_offset PAYLOAD_POSITION
+    fix_offset PRELUDE_POSITION
+  '';
+
+  dontStrip = true;
+
+  meta = rec {
+    description = "The GitHub copilot IntelliJ plugin's native component";
+    longDescription = ''
+      The GitHub copilot IntelliJ plugin's native component, patched for Nix(OS) compatibility.
```
Comment (SuperSandro2000) [inline pkgs/development/tools/github-copilot-intellij-agent/default.nix:65]:
```suggestion
      The GitHub copilot IntelliJ plugin's native component.
```

File: pkgs/development/tools/github-copilot-intellij-agent/default.nix [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      bin/copilot-agent must be symlinked into the plugin directory, replacing the existing binary.
+
+      For example:
+
+      ```shell
+      ln -s /run/current-system/sw/bin/copilot-agent ~/.local/share/JetBrains/IntelliJIdea2022.2/github-copilot-intellij/copilot-agent/bin/copilot-agent-linux
```
Comment (SuperSandro2000) [inline pkgs/development/tools/github-copilot-intellij-agent/default.nix:71]:
```suggestion
      ln -fs /run/current-system/sw/bin/copilot-agent ~/.local/share/JetBrains/IntelliJIdea2022.2/github-copilot-intellij/copilot-agent/bin/copilot-agent-linux
```

the hardcoded version number is not that nice. Is there no better way?

File: pkgs/development/tools/github-copilot-intellij-agent/default.nix [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+    patchelf --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" ${binaryLocation}
+    patchelf --set-rpath ${libPath} ${binaryLocation}
+    chmod +x ${binaryLocation}
+
+    new_size=$(stat --printf=%s ${binaryLocation})
+
+    var_skip=20
+    var_select=22
+    shift_by=$(expr $new_size - $orig_size)
```
Comment (SuperSandro2000) [inline pkgs/development/tools/github-copilot-intellij-agent/default.nix:45]: I expect this to potential break with an update. Am I correct in that?

File: pkgs/development/tools/github-copilot-intellij-agent/default.nix [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+stdenv.mkDerivation rec {
+  pname = "github-copilot-intellij-agent";
+  version = "1.2.8.2631";
+
+  src = let updateId = 341846;
+  in fetchurl {
+    name = "${pname}-${version}-plugin.zip";
+    url = "https://plugins.jetbrains.com/plugin/download?updateId=${toString updateId}";
+    hash = "sha256-0nnSMdx9Vb2WyNHreOJMP15K1+AII/kCEAOiFK5Mhik=";
+  };
```
Comment (SuperSandro2000) [inline pkgs/development/tools/github-copilot-intellij-agent/default.nix:12]:
```suggestion
  src = fetchurl {
    name = "${pname}-${version}-plugin.zip";
    url = "https://plugins.jetbrains.com/plugin/download?updateId=341846";
    hash = "sha256-0nnSMdx9Vb2WyNHreOJMP15K1+AII/kCEAOiFK5Mhik=";
  };
```

File: pkgs/development/tools/github-copilot-intellij-agent/default.nix (hunk 3 above)
Comment (hacker1024) [reply pkgs/development/tools/github-copilot-intellij-agent/default.nix:45]: I don't think so. Those values are unchanged from the original author's use in another derivation, and I have updated this one several times without anything breaking.

File: pkgs/development/tools/github-copilot-intellij-agent/default.nix (hunk 2 above)
Comment (hacker1024) [reply pkgs/development/tools/github-copilot-intellij-agent/default.nix:71]: Personally, I use Home Manager along with the IDE derivation `version` values.
I think this is too complex for the example, though.

```nix
xdg.dataFile =
  let
    ideDataDirectories = with pkgs;
      with jetbrains; {
        "JetBrains/IntelliJIdea" = idea-ultimate;
        "JetBrains/PyCharm" = pycharm-professional;
        "JetBrains/WebStorm" = webstorm;
        "JetBrains/CLion" = clion;
        "Google/AndroidStudioPreview" = androidStudioPackages.dev;
      };
    copilotAgent = "${github-copilot-intellij-agent}/bin/copilot-agent";
  in
  lib.mapAttrs'
    (name: package:
      lib.nameValuePair "${name}${
  lib.versions.majorMinor (lib.getVersion package)
}/github-copilot-intellij/copilot-agent/bin/copilot-agent-linux"
        {
          source = copilotAgent;
          force = true;
        })
    ideDataDirectories;
```

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**